File: perl/odometer.pl
#!/usr/bin/perl

=head1 NAME

odometer.pl - filters strings of digits into alphanumeric string via the T9 telephone keypad

=head1 SYNOPSIS

    usage: [perl] odometer.pl

=head1 DESCRIPTION

Read strings of digits from STDIN, one line at a time.
Convert each string of digits into a set of alphanumeric strings via the T9 telephone keypand
using odometer algorithm and write the set of results to STDOUT.

=cut

use strict;
use warnings;

our %T9 = (
    '0' => [ '0' ],
    '1' => [ '1' ],
    '2' => [ 'A', 'B', 'C' ],
    '3' => [ 'D', 'E', 'F' ],
    '4' => [ 'G', 'H', 'I' ],
    '5' => [ 'J', 'K', 'L' ],
    '6' => [ 'M', 'N', 'O' ],
    '7' => [ 'P', 'Q', 'R', 'S' ],
    '8' => [ 'T', 'U', 'V' ],
    '9' => [ 'W', 'X', 'Y', 'Z' ]
);
our $VERSION = '1.0.0';

###########################################################################
###########################################################################
# Generate strings by constructing a kind of odometer.
sub odometer (_) {
    my  $digits = $_;
    my  @wheels = ( );
    
    # Build the 'odometer' with different wheels at each position
    # based on the chars mapped from the original string.
    for (my $i = 0; $i < length($digits); $i++) {
        my  $focus = substr($digits, $i, 1);
        my  $chars = $T9{$focus} || [ $focus ];
        
        $wheels[$i] = {
            click => 0,
            chars => $chars
        };
    }
    
    my  @result = ( );
    my  $which  = 0;
    
    while ($which >= 0) {
        # Next string is from current state of 'odometer'.
        push @result, join '', map { $_->{chars}->[$_->{click}] } @wheels;
        
        # Start clicking wheels on the right.
        $which = scalar(@wheels) - 1;
        
        while ($which >= 0) {
            my  $wheel = $wheels[$which];
            
            $wheel->{click}++;
            
            last # ready to push out the next one, no overflow
                if $wheel->{click} < scalar @{$wheel->{chars}};
            
            # Zero this wheel and focus on the one to the left.
            $wheel->{click} = 0;
            $which--;
        }
    }
    
    @result
}

###########################################################################
###########################################################################

my  $app = $0 =~ m|.*/(.*)| ? $1 : 'odometer.pl';

warn "### Starting $app $VERSION\n";

while (<STDIN>) {
    chomp;
    
    my  @result = odometer $_;
    
    printf "%-25s%7d result%s\n", $_, scalar @result, @result == 1 ? '' : 's';
    
    my  $count = 79 / (length($_) + 1);
    
    while (@result) {
        print ' ';
    
        for (my $i = 0; $i < $count; $i++) {
            last
                unless @result;
            
            print ' ' . shift @result;
        }
        
        print "\n";
    }
}

warn "### Finished $app $VERSION\n";

###########################################################################
###########################################################################

=head1 AUTHOR

Marc M. Adkins <Perl-at-Doorways-dot-org>

=head1 COPYRIGHT

Copyright 2019 by Marc M. Adkins

=cut

File: dart/bin/odometer.dart
import 'dart:io';

var t9 = {
  '0': ['0'],
  '1': ['1'],
  '2': ['A', 'B', 'C'],
  '3': ['D', 'E', 'F'],
  '4': ['G', 'H', 'I'],
  '5': ['J', 'K', 'L'],
  '6': ['M', 'N', 'O'],
  '7': ['P', 'Q', 'R', 'S'],
  '8': ['T', 'U', 'V'],
  '9': ['W', 'X', 'Y', 'Z'],
};

const version = '1.0.0';

class Wheel {
  Wheel(this.chars);
  int position = 0;
  List<String> chars;

  bool click() {
    if (++position < chars.length) {
      return false;
    } else {
      position = 0;
      return true;
    }
  }

  String get current => chars[position];
}

List<String> odometer(String digits) {
  List<Wheel> wheels = List.generate(digits.length, (i) {
    var focus = digits[i];
    var chars = t9[focus];
    chars ??= [focus];
    return Wheel(chars);
  });

  List<String> result = List.empty(growable: true);

  while (true) {
    //  Next result string is from current state of 'odometer'.
    result.add(
        List.generate(wheels.length, (index) => wheels[index].current).join());

    var carryOver = true;
    for (var wheel in wheels.reversed) {
      carryOver = wheel.click();
      if (!carryOver) {
        break;
      }
    }
    if (carryOver) {
      return result;
    }
  }
}

void main(List<String> arguments) {
  stderr.write("### Starting odometer.go $version\n");

  var line = stdin.readLineSync();
  while (line != null) {
    var digits = line.trim();
    var results = odometer(digits);
    var resultWord = results.length == 1 ? "result" : "results";

    // Dart has no sprintf for some reason (though it's available as a 3rd part package).
    // The following uses the Dart string interpolation mechanism just to be Dart-ish.
    stdout.write(
        "${digits.padRight(25)}${results.length.toString().padLeft(7, " ")} $resultWord\n");

    var count = 79 ~/ (digits.length + 1);
    var r = 0;
    while (r < results.length) {
      stdout.write(' ');
      for (var i = 0; i <= count; i++) {
        if (r < results.length) {
          stdout.write(" ${results[r++]}");
        } else {
          break;
        }
      }
      stdout.write("\n");
    }
    line = stdin.readLineSync();
  }

  stderr.write("### Finished odometer.go $version\n");
}
